### PR TITLE
global: disable THP for Ceph daemons

### PR DESCRIPTION
**Pull request by batrick**
I've tested this on vstart and observed the correct prctl calls with strace. It should work as advertised.

### CODE REVIEW

**Comment by alfredodeza** [comment]
Why is this an option if we are discouraging it?

**Comment by batrick** [comment]
> Why is this an option if we are discouraging it?

It can be enabled when doing performance testing without changing the default behavior. The `prctl` cannot be overridden through `madvise` calls.

**Comment by jdurgin** [comment]
Is it possible for other processes on the system using huge page to consume enough memory with huge pages to starve the ceph processes?

File: src/common/options.cc [hunk 1]
```diff
@@ -786,6 +786,12 @@ std::vector<Option> get_global_options() {
     .set_flag(Option::FLAG_NO_MON_UPDATE)
     .set_description(""),
 
+    Option("thp", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
```
Comment (jdurgin) [inline src/common/options.cc]: might as well make this a dev option, no reason a regular user should ever change this

**Comment by markhpc** [comment]
@jdurgin Assuming that those processes are running in containers, I would think they would hit the same OOM conditions that Ceph is hitting if the cgroup's total_rss exceeds the limit that is set? IE I don't think it would necessarily hurt us, just those other applications may themselves be OOM killed?

**Comment by liewegas** [comment]
> I'm not in love with this approach since I don't think we should be taking on responsibility for overriding a system-wide setting that tells users that THP is "always" on. I see it as sort of a last resort on operating systems that absolutely refuse to set THP's default to "madvise" giving applications the choice to opt-in where appropriate. Having said that, it's better right now for users if Ceph avoids THP so I'm reluctantly approving this. In the long run we should probably work toward memory allocation behavior that utilizes THP when beneficial with MADV_HUGEPAGE.

I share the frustration with the kernel's default behavior, but the fact remains that it behaves this way today, and we need to work around it unless/until that behavior is changed.

**Comment by markhpc** [comment]
> > I'm not in love with this approach since I don't think we should be taking on responsibility for overriding a system-wide setting that tells users that THP is "always" on. I see it as sort of a last resort on operating systems that absolutely refuse to set THP's default to "madvise" giving applications the choice to opt-in where appropriate. Having said that, it's better right now for users if Ceph avoids THP so I'm reluctantly approving this. In the long run we should probably work toward memory allocation behavior that utilizes THP when beneficial with MADV_HUGEPAGE.
> 
> I share the frustration with the kernel's default behavior, but the fact remains that it behaves this way today, and we need to work around it unless/until that behavior is changed.

I generally agree, which is why I'm reluctantly approving this.  Just to make sure we aren't wrongly blaming the kernel guys though, the kernel default behavior is actually [madvise]:

```
"madvise" will enter direct reclaim like "always" but only for regions
that are have used madvise(MADV_HUGEPAGE). This is the default behaviour.
```
https://www.kernel.org/doc/Documentation/vm/transhuge.txt

It's certain operating systems that are changing the default behavior to [always].  I fear that this change is going to encourage them to simply continue overriding the kernel default.  Nevertheless, I agree we probably have to do it if we can't get those vendors to change their behavior.

For reference, the Ubuntu team actually did testing on this and saw that by switching to the kernel default (ie madvise) they saw a 1-7% improvement 1-512M stream tests and a 13-30% improvement in malloc/free tests.

https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1703742/comments/10

File: src/common/options.cc (hunk 1 above)
Comment (batrick) [reply src/common/options.cc]: done

**Comment by batrick** [comment]
> Is it possible for other processes on the system using huge page to consume enough memory with huge pages to starve the ceph processes?

In a hyperconverged setup with memory reservations, probably not? I'm not certainn.

**Comment by batrick** [comment]
DNM to make some amendments to the commit message; it's fine to QA this.

**Comment by markhpc** [comment]
> > > I'm not in love with this approach since I don't think we should be taking on responsibility for overriding a system-wide setting that tells users that THP is "always" on. I see it as sort of a last resort on operating systems that absolutely refuse to set THP's default to "madvise" giving applications the choice to opt-in where appropriate. Having said that, it's better right now for users if Ceph avoids THP so I'm reluctantly approving this. In the long run we should probably work toward memory allocation behavior that utilizes THP when beneficial with MADV_HUGEPAGE.
> > 
> > 
> > I share the frustration with the kernel's default behavior, but the fact remains that it behaves this way today, and we need to work around it unless/until that behavior is changed.
> 
> I generally agree, which is why I'm reluctantly approving this. Just to make sure we aren't wrongly blaming the kernel guys though, the kernel default behavior is actually [madvise]:
> 
> ```
> "madvise" will enter direct reclaim like "always" but only for regions
> that are have used madvise(MADV_HUGEPAGE). This is the default behaviour.
> ```
> 
> https://www.kernel.org/doc/Documentation/vm/transhuge.txt
> 
> It's certain operating systems that are changing the default behavior to [always]. I fear that this change is going to encourage them to simply continue overriding the kernel default. Nevertheless, I agree we probably have to do it if we can't get those vendors to change their behavior.
> 
> For reference, the Ubuntu team actually did testing on this and saw that by switching to the kernel default (ie madvise) they saw a 1-7% improvement in 1-512M stream tests and a 13-30% improvement in malloc/free tests:
> 
> https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1703742/comments/10

Nope, I was wrong about this.  The kernel *documentation* makes it sound like madvise is the default, but in reality mm/Kconfig sets always as default so any distributions that don't explicitly set CONFIG_TRANSPARENT_HUGEPAGE_MADVISE=y will end up with always.  See:

https://github.com/torvalds/linux/blob/master/mm/Kconfig#L385-L407

**Comment by batrick** [comment]
> > > > I'm not in love with this approach since I don't think we should be taking on responsibility for overriding a system-wide setting that tells users that THP is "always" on. I see it as sort of a last resort on operating systems that absolutely refuse to set THP's default to "madvise" giving applications the choice to opt-in where appropriate. Having said that, it's better right now for users if Ceph avoids THP so I'm reluctantly approving this. In the long run we should probably work toward memory allocation behavior that utilizes THP when beneficial with MADV_HUGEPAGE.
> > > 
> > > 
> > > I share the frustration with the kernel's default behavior, but the fact remains that it behaves this way today, and we need to work around it unless/until that behavior is changed.
> > 
> > 
> > I generally agree, which is why I'm reluctantly approving this. Just to make sure we aren't wrongly blaming the kernel guys though, the kernel default behavior is actually [madvise]:
> > ```
> > "madvise" will enter direct reclaim like "always" but only for regions
> > that are have used madvise(MADV_HUGEPAGE). This is the default behaviour.
> > ```
> > 
> > 
> > https://www.kernel.org/doc/Documentation/vm/transhuge.txt
> > It's certain operating systems that are changing the default behavior to [always]. I fear that this change is going to encourage them to simply continue overriding the kernel default. Nevertheless, I agree we probably have to do it if we can't get those vendors to change their behavior.
> > For reference, the Ubuntu team actually did testing on this and saw that by switching to the kernel default (ie madvise) they saw a 1-7% improvement in 1-512M stream tests and a 13-30% improvement in malloc/free tests:
> > https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1703742/comments/10
> 
> Nope, I was wrong about this. The kernel _documentation_ makes it sound like madvise is the default, but in reality mm/Kconfig sets always as default so any distributions that don't explicitly set CONFIG_TRANSPARENT_HUGEPAGE_MADVISE=y will end up with always. See:
> 
> https://github.com/torvalds/linux/blob/master/mm/Kconfig#L385-L407

Got it. I've updated the commit message.

**Comment by batrick** [comment]
The setting is propagated via fork and even execve. I haven't done testing but I don't see a lot of reason to. QE will verify this eventually.

**Comment by vasukulkarni** [comment]
The commit message has good details which I think should also go in docs.

**Pull request merged**